### PR TITLE
Improve caching for read_marker API

### PR DESCRIPTION
**Pull request by erikjohnston**
(I've also added some sytests https://github.com/matrix-org/sytest/pull/433)

### CODE REVIEW

**Comment by erikjohnston** [comment]
The failed Sytest is failing on `If remote user leaves room we no longer receive device updates`, which feels spurious.

**Comment by erikjohnston** [comment]
The other sytest failed because perl failed to install `Alien::Sodium`

File: synapse/storage/account_data.py [hunk 1]
```diff
@@ -156,6 +157,38 @@ def get_account_data_for_room_txn(txn):
             "get_account_data_for_room", get_account_data_for_room_txn
         )
 
+    @cached(num_args=3, max_entries=5000)
+    def get_account_data_for_room_and_type(self, user_id, room_id, account_data_type):
+        """Get all the client account_data for a user for a room.
```
Comment (richvdh) [inline synapse/storage/account_data.py]: LIES

File: synapse/storage/account_data.py [hunk 2]
```diff
@@ -156,6 +157,38 @@ def get_account_data_for_room_txn(txn):
...
+
+        Args:
+            user_id(str): The user to get the account_data for.
+            room_id(str): The room to get the account_data for.
+            account_data_type (str): The account data type to get.
+        Returns:
+            A deferred dict of the room account_data for that type, or None if
```
Comment (richvdh) [inline synapse/storage/account_data.py]: not necessarily a dict?

File: synapse/storage/events.py [hunk 3]
```diff
@@ -2033,7 +2033,7 @@ def is_event_after(self, event_id1, event_id2):
         to_2, so_2 = yield self._get_event_ordering(event_id2)
         defer.returnValue((to_1, so_1) > (to_2, so_2))
 
-    @defer.inlineCallbacks
+    @cachedInlineCallbacks(max_entries=5000)
```
Comment (richvdh) [inline synapse/storage/events.py]: I'd really prefer it if we used
```
@cached()
@defer.inlineCallbacks
```

but I'm not going to argue too much about it

Comment (erikjohnston) [reply synapse/storage/events.py]: Me too, but I think this is something to be investigated outside this PR.

**Pull request merged**